Modules/_testbuffer.c

```diff
@@ -2813,28 +2813,9 @@ static struct PyMethodDef _testbuffer_functions[] = {
     {NULL, NULL}
 };
 
-static struct PyModuleDef _testbuffermodule = {
-    PyModuleDef_HEAD_INIT,
-    "_testbuffer",
-    NULL,
-    -1,
-    _testbuffer_functions,
-    NULL,
-    NULL,
-    NULL,
-    NULL
-};
-
-
-PyMODINIT_FUNC
-PyInit__testbuffer(void)
+static int
+testbuffer_exec(PyObject *m)
 {
-    PyObject *m;
-
-    m = PyModule_Create(&_testbuffermodule);
-    if (m == NULL)
-        return NULL;
-
     Py_SET_TYPE(&NDArray_Type, &PyType_Type);
     Py_INCREF(&NDArray_Type);
     PyModule_AddObject(m, "ndarray", (PyObject *)&NDArray_Type);
```

Comment:
Maybe you can use `PyModule_AddType` in here too?

Reply:
@shihai1991 yes and I guess I should convert these to heap types too

Reply:
@shihai1991 @vstinner I am trying to convert the NDArray_Type to a heap type but it has binary functions in its as_mapping that need access to the module state, e.g.:

`nd = (NDArrayObject *)ndarray_new(&NDArray_Type, NULL, NULL);`

The signature of this method is:
```
static PyObject *
ndarray_subscript(NDArrayObject *self, PyObject *key)
```

There doesn't seem to be a way to get the module in this case - so I believe converting to heap types might block until a future PEP.  Please correct me if I am wrong but I will proceed with using PyModule_AddType without heap types.

Reply:
Maybe you can take a look of victor's PR: https://github.com/python/cpython/commit/1c2fa781560608aa4be50c748d4b3f403cfa5035

Reply:
@shihai1991 That PR does not have any types with an as_mapping method defined

Reply:
If it is not possible to subclass the type, you can use PyType_GetModule().

Reply:
@vstinner I think I'm missing some information.   How can I use `PyType_GetModule()` on a method signature like:

```
static PyObject *
ndarray_subscript(NDArrayObject *self, PyObject *key)
```

There is no type here for me to access.

Reply:
Py_TYPE(self) gives you a type.

Reply:
@vstinner oh, right.  Every object struct starts with PyObject_HEAD which has the type.  I totally forgot about that.